staging/src/k8s.io/apiserver/pkg/server/options/audit.go

```diff
@@ -293,6 +296,9 @@ func (o *AuditBatchOptions) AddFlags(pluginName string, fs *pflag.FlagSet) {
 	fs.IntVar(&o.BatchConfig.ThrottleBurst, fmt.Sprintf("audit-%s-batch-throttle-burst", pluginName),
 		o.BatchConfig.ThrottleBurst, "Maximum number of requests sent at the same "+
 			"moment if ThrottleQPS was not utilized before. Only used in batch mode.")
+	fs.IntVar(&o.BatchConfig.MaxGoRoutines, fmt.Sprintf("audit-%s-batch-max-routines", pluginName),
```

Comment:
I don't think there's any reason to expose this flag for the log writer. There's no reason not to always set it to 1.

Do we even need to expose this for the webhook, or can we derive a sane value from the other params?

Reply:
> I don't think there's any reason to expose this flag for the log writer. There's no reason not to always set it to 1.

We already support qps for log writer. So it's the same here.

Reply:
>Do we even need to expose this for the webhook, or can we derive a sane value from the other params?

maybe with a follow up , we can deprecate qps and burst. I am not sure.

Reply:
agree that I'd prefer not to expose more tuning knobs than necessary.

Reply:
Actually qps burst take similar responsibility to control concurrency. Then  how we set max-go-routines from the burst?

Reply:
> There's no reason not to always set it to 1.

I agree with this.

Reply:
I don't think so, why we introduce qps for log backend? It is because we want concurrent processing events encoding.

Reply:
qps is rate, not concurrency

Reply:
```
// processEvents process the batch events in a goroutine using delegateBackend's ProcessEvents.
func (b *bufferedBackend) processEvents(events []*auditinternal.Event) {
	if len(events) == 0 {
		return
	}

	// TODO(audit): Should control the number of active goroutines
	// if one goroutine takes 5 seconds to finish, the number of goroutines can be 5 * defaultBatchThrottleQPS
	if b.throttle != nil {
		b.throttle.Accept()
	}

	b.wg.Add(1)
	go func() {
		defer b.wg.Done()
		defer runtime.HandleCrash()

		// Execute the real processing in a goroutine to keep it from blocking.
		// This lets the batching routine continue draining the queue immediately.
		b.delegateBackend.ProcessEvents(events...)
	}()
}
```

I know, the code above can start up many routines `ProcessEvents` for log backend.

Reply:
It is no difference for any delegateBackend.

Reply:
> I know, the code above can start up many routines ProcessEvents for log backend.


So we need a way to limit it to `one` for log backend?

:confused: :confused: :confused:

Reply:
@liggitt @tallclair for final aggreement. What should we do?  Why need muti routines for log backend comes after a discussion with @crassirostris long ago.

Reply:
We need multiple routines for the webhook, it doesn't make sense for the log backend. All multiple routines would do is create contention for the writer lock (decrease performance) and cause lines to be out of order. QPS & burst make sense for the log backend because they can limit the amount of disk pressure created by the audit log.

For the webhook, the difference between QPS & burst is that those settings impact the external server (where the logs get pushed to), while the number of concurrent routines only matters on the apiserver host, as a way of keeping memory bounded. What I'm wondering is, is there a sane default we can for this rather than making the user figure it out for themselves? E.g. if we expect the server to always respond in <5s, then the maximum routines should be 5 * (qps + burst), maybe with extra wiggle room for retries.

Reply:
clearly, thanks @tallclair

Reply:
@tallclair I think you mean 5 * qps

reworking on this issue: 
I found i do not say clearly, the bufferedBackend run a routine collect and then process events(`processEvents`) from its buffer.
`processEvents` is where to startup many goroutines calling `delegateBackend.ProcessEvents`.

Buffered backend does not care which delegate backend is.  If delegate is log backend. We call `logbackend.ProcessEvents` in many routines just as what we are doing in `staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go`. 

So this below I think you mean the lower writer.  Other similar example, we can call glog.Info concurrently without logs mixed.
>I don't think there's any reason to expose this flag for the log writer. There's no reason not to always set it to 1.